### ISSUE TITLE
[perf] Unnecessary memory growth check repetitions

### ISSUE THREAD

**Issue by RReverser** [emscripten-core/emscripten#11059]
For background: I've been looking through https://github.com/WebAssembly/design/issues/1271 referenced from console warnings when building Wasm with threads, as well as implementation PR.

I think current Emscripten's implementation of tracking memory growth has implementation details that make it much slower than it has to be and could be made much faster even Wasm API design aside.

For example, I've looked at the generated JS of the benchmark from https://github.com/WebAssembly/design/issues/1271#issuecomment-477257958 and its performance profiles, and noticed that it's spending quite a lot of time in `GROWABLE_HEAP_*` functions.

On its own it's unsurprising, but upon closer look it seems to be not so much due to the slowness of functions themselves (where the only expensive part is the `wasmMemory.buffer` accessor) but rather due to number of times they are called, which significantly multiplies that constant-time cost.

For example, here is what the generated `AsciiToString` - one of the highest callees - looks like:

```js
function AsciiToString(ptr) {
 var str = "";
 while (1) {
  var ch = GROWABLE_HEAP_U8()[ptr++ >> 0];
  if (!ch) return str;
  str += String.fromCharCode(ch);
 }
}
```

Since it's expected to read an existing string, there is no reason to repeatedly check for memory growth for every single byte - the view can't get invalidated anyway, and we don't care about higher bytes even if memory grows.

Changing this single function to

```js
function AsciiToString(ptr) {
 var str = "";
 var heapU8 = GROWABLE_HEAP_U8();
 while (1) {
  var ch = heapU8[ptr++ >> 0];
  if (!ch) return str;
  str += String.fromCharCode(ch);
 }
}
```

made the benchmark on my machine in latest stable Chrome go from 5.7s to 4.1s - 28% improvement.

There seems to be a lot of other functions (e.g. `readAsmConstArgs`) that do the same repetitive checks inside a loop, presumably due to the templating Emscripten uses for the `HEAP` variables that substitutes all such accessors with function calls.

Fixing few more suspects quickly brought performance down to 2.9s - almost 50% of the original time.

I don't know if it's possible to automatically check+update memory views only on edges where Wasm calls / returns to JS (e.g. by instrumenting imports + ASM_CONSTS), or whether this is something that needs to be changed manually on per-function basis, but I thought I'd post this anyway, as fixing such loops seems to be a very low-hanging fruit that could significantly improve performance of pthread + memory growth builds.

**Comment by kripken** [comment]
Interesting, I'm disappointed the JS VM didn't do better, as if it's a hot loop it should inline it and be able to remove much of the overhead. But I guess not all.

To optimize this, two options might be

1. Hand-optimize such functions. The optimizer would then need to look for any such hand-optimization, namely, one of the `GROWABLE_*`, and if it sees that, it can skip the function.
2. Add some acorn optimizer passes to find loops without unsafe calls, like `AsciiToString`, and hoist the heap functions.

The bigger question for me though is, how important is this? Pthreads + growth is a rare combination in general, and that + heavy `AsciiToString` or such is likely even more rare.

**Comment by RReverser** [comment]
> I'm disappointed the JS VM didn't do better, as if it's a hot loop it should inline it and be able to remove much of the overhead

I think the problem here is that `.buffer` is a non-trivial getter - it actually needs to go into C++ code, and read a value from memory that is also shared between different threads.

I don't think even native compiler could optimise this away, since it's not clear that the user doesn't *actually* need up-to-date buffer on each iteration (in case the memory got resized by a different thread meanwhile).

> Pthreads + growth is a rare combination in general

Yes, but could it be rare specifically due to the noticeable warning about [also potentially noticeable] performance problems? Seems like chicken-and-egg problem. I'd imagine most apps actually do want memory growth to accomodate for arbitrary user input.

> To optimize this, two options might be

I'm not sure what you mean by "unsafe calls", but, as mentioned above, I was wondering if 3rd option - updating memory locations on JS<->Wasm crossings could also be an option. Are there any functions that actually need an up-to-date `.buffer` on loop iteration? If not, then we could probably wrap imports/exports to update buffers only at such call boundary.

**Comment by kripken** [comment]
> I was wondering if 3rd option - updating memory locations on JS<->Wasm crossings could also be an option.

I think that might not be good enough. Imagine a JS loop on a linked list that is modified by wasm on another thread. That's not likely, I agree that normally a JS function would get a pointer and just use that one object. But if it chases pointers, it may end up needing a new buffer.

**Comment by RReverser** [comment]
> Imagine a JS loop on a linked list that is modified by wasm on another thread

That's what I'm asking - does Emscripten have any built-in JS functions that do something weird like that? Note that we don't care about user-defined functions like `EM_JS`, since user JS has control over accessing the memory anyway and can choose whether they want to use the variable or call the `GROWABLE_*` function.

**Comment by kripken** [comment]
I think user code is relevant here. We automatically instrument user code to made it call `GROWABLE_*`, exactly like we instrument our builtins. That makes it safe by default. (We could perhaps allow users to opt out of that, if they are sure they are doing the right thing themselves?)

If this type of optimization is just for builtins, then we can hand-optimize them probably (option 1). You're right, none of them do anything weird like that.

**Comment by RReverser** [comment]
> We automatically instrument user code to made it call GROWABLE_*, exactly like we instrument our builtins.

Ah, TIL. Do you mean variables (`HEAP*`) are replaced with such calls? Yeah, it probably will require a built-in. Anyway, I think in that case we can just leave such instrumentation for now.

> then we can hand-optimize them probably (option 1). You're right, none of them do anything weird like that.

The difference is in how many times buffers are updated per JS call.

An example: if a single `EM_JS` function calls `AsciiToString` on some array of pointers, then instrumenting imports/exports is more benefitial, because it would allow to update buffers once at crossing the boundary and then reusing same updated `HEAPU8` variable in all `AsciiToString` invocation. But if we hand-optimise each built-in, then we still have at least one `GROWABLE_*` call per each string.

In most cases the difference might not matter, but if we add some caching anyway, why not to it in slightly more efficient way?

**Comment by sbc100** [comment]
Would it help to add UNCHECKED_HEAPXX for cases where people know they have a valid pointer and want to avoid the check?

**Comment by kripken** [comment]
> But if we hand-optimise each built-in, then we still have at least one GROWABLE_* call per each string.

True. But one such call per string is probably very low overhead (unless the strings are super-short). On the other hand, if there's a simple way to optimize at the import/export level, I'd be happy with that.

> Would it help to add UNCHECKED_HEAPXX for cases where people know they have a valid pointer and want to avoid the check?

And that would assume the heap was updated on crossing the import/export boundary? Yeah, sounds like that could work.

In general though my bigger question, as I said earlier, is I'm unsure how important this is. Do we have real-world use cases that need to be sped up?

**Comment by RReverser** [comment]
Bump to avoid closing, we should probably find real-world use-cases though. We could use Squoosh, where we use multithreading and memory growth, but we don't cross into JS frequently there so this likely doesn't affect us much.
